### PR TITLE
root_path added to build config and file locations

### PR DESCRIPTION
**Pull request by ServerlessSam**
`root_path` is added as a property of `BuildConfig` and this is fed into all uses of `FileLocation`.

Also added a new flag to the CLI. If present, the `root_path` is set as the value following the flag, if not, the `DFM_ROOT_PATH` env var is used. If neither are present, the root path defaults to the root of the PC's drive (e.g c:// or / based on OS)

### CODE REVIEW

File: src/cli.py [hunk 1]
```diff
@@ -20,6 +23,19 @@ def parse_parameter_string(param_str: str) -> dict:
...
+def get_root_path_from_env_var(env_var_name: str) -> Path:
+    os_default_root_path_mapping = {
+        "Windows": "c://",
+        "Darwin": "/",
+        "Linux": "/",
+        "Java": "/",
+        "": "/",
+    }
+    return Path(
+        os.getenv(env_var_name, os_default_root_path_mapping[platform.system()])
```
Comment (james-ecd) [inline src/cli.py:35]: could default the dict call here for safety:
```suggestion
        os.getenv(env_var_name, os_default_root_path_mapping.get(platform.system(), "/"))
```

File: src/cli.py [hunk 2]
```diff
@@ -49,21 +65,35 @@ def main():
...
     parser.add_argument(
         "--version",
         action="version",
         version="%(prog)s {version}".format(version=__version__),
     )
     args = parser.parse_args()
-    print(args)
+
+    # Determine root path
+    root_path = None
```
Comment (james-ecd) [inline src/cli.py:82]: dont need this line

File: src/cli.py [hunk 3]
```diff
@@ -49,21 +65,35 @@ def main():
...
+    if args.root_path:
+        root_path = Path(args.root_path)
+    else:
+        root_path = get_root_path_from_env_var("DFM_ROOT_PATH")
```
Comment (james-ecd) [inline src/cli.py:86]:
```suggestion
        root_path = Path(args.root_path) if args.root_path else get_root_path_from_env_var("DFM_ROOT_PATH")
```

File: src/cli.py [hunk 4]
```diff
@@ -49,21 +65,35 @@ def main():
...
 
     # Parse parameters
     if args.parameters:
         parameters = parse_parameter_string(args.parameters)
     else:
         parameters = None
     if args.action == "merge":
```
Comment (james-ecd) [inline src/cli.py:93]: could be a good situation for https://learnpython.com/blog/python-match-case-statement/

File: src/config.py [hunk 5]
```diff
@@ -63,10 +62,12 @@ def __post_init__(self):
     def file_content(self) -> dict | list:
         return (
             JsonFileType.load_from_file(
-                Path(ROOT_PATH) / self.destination_file_location.substituted_path
+                self.destination_file_location.root_path
+                / self.destination_file_location.substituted_path
```
Comment (james-ecd) [inline src/config.py:66]: confused how this works

File: src/cli.py (hunk 4 above)
Comment (ServerlessSam) [reply src/cli.py:92]: Will go with your suggested one-liner for if else instead

File: src/config.py (hunk 5 above)
Comment (ServerlessSam) [reply src/config.py:66]: I have added `root_path` as an attribute of `FileLocation`. This class has the `destination_file_location` attribute of type `FileLocation`. So from this attribute we obtain the root path. The same attribute also has fully resolved (and potentially substituted) paths.

However as I write this, that is a relative path FROM the root path. E.g root path of `/foo` and substituted path of `bar/baz.json` tell the program the file is at `/foo/bar/baz.json`.

As I write this, I'm realising now that the `FileLocation` contains the root path, we should just set the substituted path to be full paths I think. From this point onwards in the code I think we are just constantly merging the root path and any relative paths.

What do you think?

And does this clear things up?

Comment (ServerlessSam) [reply src/config.py:66]: @james-ecd ?

Comment (james-ecd) [reply src/config.py:66]: I meant I'm not sure how the forward slashes in between those strings is valid syntax

Comment (ServerlessSam) [reply src/config.py:66]: https://stackoverflow.com/questions/48190959/how-do-i-append-a-string-to-a-path-in-python

Comment (ServerlessSam) [reply src/config.py:66]: Wagwan

**Review by james-ecd — APPROVED** [review]

**Pull request merged**